File: plugins/collector/event/apps/web/tengine.go
package web

import (
	"collector/cache/process"
	"collector/event/apps"
	"regexp"
	"strings"
)

type Tengine struct {
	version  string
	reg      *regexp.Regexp
	regNginx *regexp.Regexp
}

func (Tengine) Name() string { return "tengine" }

func (Tengine) Type() string { return "web" }

func (t Tengine) Version() string { return t.version }

// pay attention to the name, maybe nginx
func (Tengine) Match(p *process.Process) bool { return p.Name == "tegine" }

func (t *Tengine) Run(p *process.Process) (m map[string]string, err error) {
	m = make(map[string]string)
	m["nginx_version"] = ""
	if t.reg == nil {
		t.reg = regexp.MustCompile(`Tengine\/(\d+\.)+\d+`)
		t.regNginx = regexp.MustCompile(`nginx\/(\d+\.)+\d+`)
	}
	result, err := apps.Execute(p, "-v")
	str := t.reg.FindString(result)
	if str == "" {
		err = apps.ErrVersionNotFound
		return
	}
	if s := t.regNginx.FindString(result); s != "" {
		m["nginx_version"] = strings.TrimPrefix(str, "nginx/")
	}
	t.version = strings.TrimPrefix(str, "Tengine/")
	return
}

func init() {
	apps.Regist(&Tengine{})
}

File: plugins/collector/src/event/apps/web/tengine.rs
use crate::event::apps::{execute, AppProc, IApp};
use std::collections::HashMap;

pub struct Tengine {
    version: String,
}

impl Tengine {
    pub fn new() -> Self { Tengine { version: String::new() } }
}

impl IApp for Tengine {
    fn name(&self)     -> &'static str { "tengine" }
    fn app_type(&self) -> &'static str { "web" }
    fn version(&self)  -> &str         { &self.version }

    fn matches(&mut self, p: &AppProc) -> bool { p.name == "tengine" }

    fn run(&mut self, p: &AppProc) -> anyhow::Result<HashMap<String, String>> {
        let out = execute(p, &["-v"])?;
        // "Tengine/2.3.3"
        if let Some(v) = out.split_whitespace().find(|s| s.starts_with("Tengine/")) {
            self.version = v.trim_start_matches("Tengine/").to_owned();
        }
        Ok(HashMap::new())
    }
}
